### ISSUE TITLE
SSAO

### ISSUE THREAD

**Issue by Usnul** [mrdoob/three.js#4955]
The SSAO shader that we have is a lot better than having nothing at all, but it is a little on the slow side and it is not the most pleasing in terms of the result. Are there any plans to improve on it? say to get Scalable Ambient Obscurance like the one implemented here:
http://floored.com/luma_demo_ao.html
according to the paper on it (https://research.nvidia.com/sites/default/files/publications/McGuire12SAO.pdf) it's a fair bit faster than SSAO. Playing around with setting in demo you can quite clearly see that SAO gives much better quality at much lower cost, it also has parameters to adjust the quality for that performance/quality sweetspot.

here is the link to original article by floored on this:
http://floored.com/blog/2013/ssao-screen-space-ambient-occlusion.html

original paper also came with glsl 1.2 implementation for their algorithm, you can download it at:
https://research.nvidia.com/publication/scalable-ambient-obscurance

**Comment by daoshengmu** [comment]
@mrdoob . I found that you changed the renderer.addPrePlugin interface, and there is no SSAO example. Therefore, I made some modifications on my past SSAO demo. Do you need me to add this demo to the dev. version? 
![screen shot 2015-01-15 at 5 16 41 pm](https://cloud.githubusercontent.com/assets/1263177/5755031/de6de4ee-9cda-11e4-80b4-b55c5c1c19ab.png)

Moreover, I will start to do Scalable Ambient Obscurance example this month. Looking forward my good news.

**Comment by mrdoob** [comment]
Oh! If you have SSAO code that doesn't depend on .addPrePlugin that would be great to have! :)

**Comment by daoshengmu** [comment]
Yup. I didn't depend on .addPrePlugin. I just use overrideMaterial and RenderTarget.
```
scene.overrideMaterial
renderer.render( scene, camera, postprocessing.rtTextureDepth , true );
```

**Comment by mrdoob** [comment]
Nice! Feel free to do a PR for it ^^

**Comment by m1ga** [comment]
@daoshengmu any news on the ssao demo?

**Comment by daoshengmu** [comment]
@m1ga. I have done the ssao demo and working on Scalable Ambient Obscurance. I think I will finish the Scalable Ambient Obscurance demo in two weeks.

**Comment by bhouston** [comment]
PS.  I found WebGL glsl code for SAO here:

https://gist.github.com/fisch0920/6770311

Haven't had time to test it yet.

**Comment by daoshengmu** [comment]
@bhouston . Thank you. I am also implementing the approach in HPG2012.

**Comment by Mugen87** [comment]
Solved with `R99` via #15243
